src/react.ts

```diff
@@ -2,20 +2,25 @@ import { useDebugValue } from 'react'
 import { useSyncExternalStoreWithSelector } from 'use-sync-external-store/shim/with-selector'
 import createStore, {
   EqualityChecker,
-  GetState,
-  SetState,
+  Mutate,
   State,
   StateCreator,
   StateSelector,
   StoreApi,
+  StoreMutatorIdentifier,
 } from './vanilla'
 
-export function useStore<T extends State>(api: StoreApi<T>): T
-export function useStore<T extends State, U>(
-  api: StoreApi<T>,
-  selector: StateSelector<T, U>,
+export function useStore<S extends StoreApi<State>>(api: S): ExtractState<S>
+export function useStore<S extends StoreApi<State>, U>(
+  api: S,
+  selector: StateSelector<ExtractState<S>, U>,
   equalityFn?: EqualityChecker<U>
 ): U
+export function useStore<S extends StoreApi<State>>(
+  api: S,
+  selector: undefined,
+  equalityFn?: EqualityChecker<ExtractState<S>>
```

Comment:
Do we need `?` for `equalityFn`? Maybe, you want to be precise.
What do you think about adding JSDoc saying "this is supported for edge cases, but generally not recommended"?

Reply:
> Do we need `?` for `equalityFn`? Maybe, you want to be precise. That's fair.

Yeah exactly to be precise, as `useStore(undefined, undefined)` should also compile (I too thought about this but then came to conclusion this is how it should be)

> What do you think about adding JSDoc saying "this is supported for edge cases, but generally not recommended"?

I don't really follow why it would be an edge case to require something `useStore(undefined, shallowEq)`. But if you think that users should be cautious then there's no harm in adding a JSDoc. Also if you think it'd be better to remove this overload then I'm open to that too, given you know better.

Reply:
To me, it's important to have types for good docs and it doesn't have to be precise about the JS implementation. (TS users can always use type assertion if they want.)
If I were a user, I would read all three overloaded types to understand the entire API, but for almost all cases, the first two are enough. I don't see the need for the third one, because it's almost always a bad practice to use equalityFn for the entire state object. (It's a hacky optimization which can behave inconsistently.) So, it's an extra mental overhead.

If adding JSDoc works (if it shows in editors), let's do it. Otherwise, let's remove.
```js
/**
 * Caution: Although this signature is supported, it's not generally a good practice to depend on equalityFn for the entire state.
 */
```
Does JSDoc work for `useBoundStore` and `UseContextStore` too?


> require something like `useStore(undefined, shallowEq)`

We would never need this. For 99.9% of use cases, the behavior is the same as `useStore()`. Put it in another way, `equalityFn` is **designed** to be used when `selector` is specified. If I see using `equalityFn` without `selector`, I suspect it's a mis-usage for 99% of use cases.

> should also compile

Maybe, we are coming from the different world. Mine is more or less "what TS compiles should work in JS".

Reply:
> For 99.9% of use cases, the behavior is the same as useStore().

Ah right, rethinking about it this and it's true that `useStore(undefined, someEquality)` is going to be same as `useStore()` (unless users are doing some nasty mutations eg `get().foo.bar = 41` or whatever which is obviously not valid usage).

Okay so this changes my entire perspective, I was under impression that selector and equality are more or less orthogonal and the type is `(selector?: ..., equals?: ...) => ...` (much like `(options: { selector?: ..., equals?: ... }) => ...`) but it seems to be the type is `{ (): ..., (selector: ..., equals?: ...): ... }` ie custom equality truly only makes sense in presence of a selector.

So yeah let's remove this overload.

Reply:
> So yeah let's remove this overload.

👍 

> is almost a coincidental unintended by-product

true. well described.

> we don't do that because it'll affect the bundle size etc.

right. and so far I don't see many people misunderstand it.